go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+	offset++
+
+	if curRunValid {
+		numValid = 1
+		re.estimatedValuesLen += int64(len(currentRun))
+	}
+	numOutput = 1
+
+	for offset < re.inputLen {
+		value, valid := re.readValue(offset)
+		offset++
+		// new run
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			if valid {
+				re.estimatedValuesLen += int64(len(value))
+			}
+
+			currentRun = value
+			curRunValid = valid
+
+			numOutput++
+			if valid {
+				numValid++
+			}
+		}
+	}
+	return
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) PreallocOutput(ctx *exec.KernelCtx, numOutput int64, out *exec.ExecResult) {
+	runEndsBuffer := ctx.Allocate(int(numOutput) * int(SizeOf[R]()))
+	var validityBuffer *memory.Buffer
+	if len(re.inputValidity) > 0 {
+		validityBuffer = ctx.AllocateBitmap(numOutput)
+	}
+
+	valueBuffer := ctx.Allocate(int(re.estimatedValuesLen))
+	offsetsBuffer := ctx.Allocate(int(numOutput+1) * int(SizeOf[O]()))
+
+	reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), re.valueType)
+	*out = exec.ExecResult{
+		Type:   reeType,
+		Len:    re.inputLen,
+		Nulls:  0,
+		Offset: 0,
+		Children: []exec.ArraySpan{
+			{
+				Type: reeType.RunEnds(),
+				Len:  numOutput,
+			},
+			{
+				Type: reeType.Encoded(),
+				Len:  numOutput,
+			},
+		},
+	}
+
+	out.Children[0].Buffers[1].WrapBuffer(runEndsBuffer)
+	if validityBuffer != nil {
+		out.Children[1].Buffers[0].WrapBuffer(validityBuffer)
+	}
+	out.Children[1].Buffers[1].WrapBuffer(offsetsBuffer)
+	out.Children[1].Buffers[2].WrapBuffer(valueBuffer)
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) WriteEncodedRuns(out *exec.ExecResult) int64 {
+	outputRunEnds := exec.GetData[R](out.Children[0].Buffers[1].Buf)
+	outputOffsets := exec.GetSpanOffsets[O](&out.Children[1], 1)
+	outputValues := out.Children[1].Buffers[2].Buf
+
+	var readOffset int64
+	currentRun, curRunValid := re.readValue(readOffset)
+	readOffset++
+
+	var writeOffset, valueOffset int64
+	validityBuf := out.Children[1].Buffers[0].Buf
+	setValidity := func(valid bool) {}
+	if len(validityBuf) > 0 {
+		setValidity = func(valid bool) {
+			bitutil.SetBitTo(validityBuf, int(writeOffset), valid)
+		}
+	}
+
+	outputOffsets[0], outputOffsets = 0, outputOffsets[1:]
+
+	writeValue := func(valid bool, value []byte) {
+		setValidity(valid)
+		valueOffset += int64(copy(outputValues[valueOffset:], value))
+		outputOffsets[writeOffset] = O(valueOffset)
+	}
+
+	for readOffset < re.inputLen {
+		value, valid := re.readValue(readOffset)
+
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			writeValue(curRunValid, currentRun)
+			runEnd := R(readOffset)
+			outputRunEnds[writeOffset] = runEnd
+			writeOffset++
+			curRunValid, currentRun = valid, value
+		}
+		readOffset++
+	}
+
+	writeValue(curRunValid, currentRun)
+	outputRunEnds[writeOffset] = R(re.inputLen)
+	return writeOffset + 1
+}
+
+func validateRunEndType[R RunEndsType](length int64) error {
+	runEndMax := MaxOf[R]()
+	if length > int64(runEndMax) {
```

Comment:
In my PR I also added a check against negative values after you pointed out this could also be an issue.

Reply:
good point! i'll add that